### ISSUE TITLE
preserve_minion_cache is broken in 2016.3+

### ISSUE THREAD

**Issue by junovitch** [saltstack/salt#35840]
### Description of Issue/Question

I use 'preserve_minion_cache: True' to maintain some files between builds.  The funtionality is broken in 2016.3.2.
### Setup

Set 'preserve_minion_cache: True'
### Steps to Reproduce Issue

Run salt-key -d <host> with 'preserve_minion_cache: True'
### Versions Report

I'm also running on CentOS.

```
Salt Version:
           Salt: 2016.3.2

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.0
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.7
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.12 (default, Jul  3 2016, 19:48:26)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.4.1
            ZMQ: 4.1.5

System Versions:
           dist:   
        machine: amd64
        release: 11.0-RC1
         system: FreeBSD
        version: Not Installed
```

**Comment by junovitch** [comment]
So the cause of this issue is https://github.com/saltstack/salt/pull/29434/commits/661f5686bf1090554d571a6ed23f09d511f5a15a.

Before that commit, we would enter https://github.com/saltstack/salt/blob/2016.3/salt/key.py#L597 with 'if not True or not ['hostname']:' when trying to do a `salt-key -d <hostname>`.  The hostname prevented the rmdir from running.  After 661f56 the condition basically becomes 'if not True or not []:' and the rmdir is run unconditionally.

I'm trying to understand the intent of what 661f56 was trying to solve.  The documentation covers the preserve_minion_cache variable but that commit references preserve_minions which does not appear to be user configurable and the change breaks the logic for the minion cache.

Additionally, 661f56 would be incomplete anyway since the RAET key is still using the old logic https://github.com/junovitch/salt/blob/2016.3/salt/key.py#L1331

A revert of 661f56 should fix the issue.

**Comment by Ch3LL** [comment]
@junovitch looks like I am able to replicate this behavior so this is definitely a regression. I'm not sure what that fix was exactly for. ping @pzipoy can you explain what the fix was for because it looks like this has caused a regression. Thanks

**Comment by junovitch** [comment]
Thank you.  I can confirm on one of my production CentOS 7 Saltmasters that reverting https://github.com/saltstack/salt/pull/29434/commits/661f5686bf1090554d571a6ed23f09d511f5a15a fixes the behavior.  I'm more than willing to submit a pull request but wanted to understand why that change was needed first before submitting a revert.

**Comment by pzipoy** [comment]
Hi @junovitch. To provide an explanation, I posted a comment regarding the original intent of pull request to remove master cache if preserve_minions is false on the original thread #29434.

**Comment by junovitch** [comment]
Additional comments added to #29434

**Comment by junovitch** [comment]
Hi, this functionality is now broken in 3 releases. I've developed a solution in https://github.com/saltstack/salt/pull/42785 that I would like to get feedback on. There are some tiny variations in 2016.11 and 2017.7 that prevent it from being a clean cherry-pick but I've started with 2016.3 per the "...the oldest release branch that contains the bug..." per the contributing guide. I'd like to get feedback on the approach and I'll follow up with vetting the fixes in 2016.11 and 2017.7. Ping @cachedout as the original committer of the functionality.
